### ISSUE TITLE
VTM and Velocity theme can’t be loaded

### ISSUE THREAD

**Issue by devemux86** [cgeo/cgeo#15035]
### Describe your problem!

There is a report in OpenAndroMaps forum about c:geo and Velocity theme not working:
https://www.openandromaps.org/en/oam-forums/topic/voluntary-uk/page/13#post-55135

I tested Voluntary / Velocity themes with Cruiser (latest VTM) and they seem to work fine.

### How to reproduce?

Use Velocity theme with VTM map engine.

### Reproducible

Yes

### c:geo Version

nightly

**Comment by eddiemuc** [comment]
Copy of the stack trace reported in above form entry for reference:

```
[main] render theme invalid
org.oscim.theme.IRenderTheme$ThemeException: width and height must be > 0
at org.oscim.theme.XmlThemeBuilder.read(XmlThemeBuilder.java:121)
at org.oscim.theme.ThemeLoader.load(ThemeLoader.java:49)
at org.oscim.theme.ThemeLoader.load(ThemeLoader.java:45)
at org.oscim.map.Map.setTheme(Map.java:197)
at org.oscim.map.Map.setTheme(Map.java:189)
at cgeo.geocaching.unifiedmap.mapsforgevtm.MapsforgeThemeHelper.reapplyMapTheme(MapsforgeThemeHelper.java:157)


```

**Comment by devemux86** [comment]
VTM [does not allow](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/theme/XmlThemeBuilder.java#L642) dash array < 1 because of OpenGL restrictions.

Also this error message does not seem exist (?) in VTM code.
So how this could happen?

Unless there is an external scale from the app on the theme rule values?

**Comment by eddiemuc** [comment]
> Also this error message does not seem to exist (?) in VTM code. So how this could happen?

Discovered the same in parallel. Message does also not exist in c:geo code. This is really strange...

@devemux86 to debug similar problems in the future it might help if ThemeException contains also the causing exception (if it is created from such an exception). Would you accept a Pull Request against vtm enabling this? E.g.:

```
try {
	... do rendering stuff
} catch (Exception ex) {
	throw new ThemeException(e.getMessage(), ex);
} finally {
	.. do closing stuff
}
```

instead of

```
try {
	... do rendering stuff
} catch (Exception ex) {
	throw new ThemeException(e.getMessage());
} finally {
	.. do closing stuff
}
```

> Unless there is an external scale from the app on the theme rule values?

There might be, but that would still not explain why the error message is nowhere to be found.

**Comment by devemux86** [comment]
@eddiemuc this is a nice idea.

Do you mean like this? https://github.com/mapsforge/vtm/pull/1083

Mapsforge should already be ok?

**Comment by eddiemuc** [comment]
> @eddiemuc this is a nice idea.
>
> Do you mean like this? [mapsforge/vtm#1083](https://github.com/mapsforge/vtm/pull/1083)

Exactly! Thanks a lot!

> Mapsforge should already be ok?

Phew, would have to look that one up....

**Comment by eddiemuc** [comment]
Meanwhile I tried to reproduce the error in c:geo using VTM and OpenAndroMaps with Velocity theme. Seems to work w/o exception:

![image](https://github.com/cgeo/cgeo/assets/6909759/9f43f0cb-d37d-4579-b764-1bccbb33cb2b)

![image](https://github.com/cgeo/cgeo/assets/6909759/2f7ec895-c322-49f2-a8fe-5defa542c98f)

![image](https://github.com/cgeo/cgeo/assets/6909759/d9c11ab9-d698-4855-bb6a-3772f622e2b3)

...except that text/symbol scale factors seems not to be applied yet. But that could be a yet-missing feature of c:geo's VTM implementation

**Comment by fm-sys** [comment]
> ...except that text/symbol scale factors seems not to be applied yet. But that could be a yet-missing feature of c:geo's VTM implementation

They are working but you need to restart cgeo to apply them. There actually should be a third one, however it was causing problems. You can read more about it here: #13593

**Comment by eddiemuc** [comment]
> > ...except that text/symbol scale factors seems not to be applied yet. But that could be a yet-missing feature of c:geo's VTM implementation
> 
> They are working but you need to restart cgeo to apply them. There actually should be a third one, however it was causing problems. You can read more about it here: #13593

Restart did not do anything. I applied a text scale of 450% but text sizes are still pretty "normal". (using openandromaps with velocity theme). I didn't try any other maps/themes. Probably there is a bug hidden here?

**Comment by moving-bits** [comment]
> > > ...except that text/symbol scale factors seems not to be applied yet. But that could be a yet-missing feature of c:geo's VTM implementation
> > 
> > 
> > They are working but you need to restart cgeo to apply them. There actually should be a third one, however it was causing problems. You can read more about it here: #13593
> 
> Restart did not do anything. I applied a text scale of 450% but text sizes are still pretty "normal". (using openandromaps with velocity theme). I didn't try any other maps/themes. Probably there is a bug hidden here?

Did a few more tests and was able to get scaling with all themes installed in this emulator (Default, Elevate/Elements, FZK and Velocity/Voluntary), although the behavior was NOT consistent: I had to restart the map sometimes, and/or open and close theme options menu (without changing anything in between). I'll create a separate issue for this.

**Comment by moving-bits** [comment]
Scaling options for VTM are fixed meanwhile. 

Does the original issue (Velocity theme does not load) still persist, or can we close here? (Have found no further issue in the related forum entry.)

**Comment by moving-bits** [comment]
No further mentions of any related problems, therefore closing here for now. Feel free to reopen, if this issue happens to strike us again.
